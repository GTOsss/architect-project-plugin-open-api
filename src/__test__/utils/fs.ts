import { parse } from 'yaml';
import { readFileSync } from 'fs-extra';

/**
 * Import yaml file as object
 * @example
 * readYaml(resolve(__dirname, '../../mock/openapi.mock.yaml'))
 * */
export const readYaml = (path: string) => {
  return parse(readFileSync(path, 'utf8'));
};
