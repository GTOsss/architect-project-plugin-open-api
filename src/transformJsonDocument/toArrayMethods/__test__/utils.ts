import { OpenAPIV3 } from 'openapi-types';
import { toArrayMethods } from '../toArrayMethods';
import { Method } from '../toArrayMethods.types';
import { TypesMap } from '../../types';

const typesMap: TypesMap = {
  integer: 'number',
  number: 'number',
  string: 'string',
  boolean: 'boolean',
  anyObject: 'AnyObject', // если в scheme только type: object
  '*': 'any', // any если не удалось определить тип
};

/**
 * Находит объект указанного метода.
 * @example findMethod(HttpMethods.GET, 'users/{id}', methods);
 * */
export const findMethod = (method: OpenAPIV3.HttpMethods, path: string, methods: Method[]): Method => {
  return methods.find((currentMethod) => currentMethod.method === method && currentMethod.path === path);
};

/**
 * Вызывает getMethodsAsArray и находит в полученном массиве указанный метод.
 * */
export const testFor = (method: OpenAPIV3.HttpMethods, path: string, jsonDocument: OpenAPIV3.Document): Method => {
  const methods = toArrayMethods(jsonDocument, { typesMap });
  return findMethod(method, path, methods);
};
