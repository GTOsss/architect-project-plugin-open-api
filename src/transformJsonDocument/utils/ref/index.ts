export * from './resolveRef';
export * from './getSchemaNameByRef';
export * from './parseRefPath';
