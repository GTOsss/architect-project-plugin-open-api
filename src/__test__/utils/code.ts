import { existsSync } from 'fs';
import { ESLint } from 'eslint';

type ApplyEslintAutoFixParams = {
  code: string;
  eslintConfigPath: string;
  filePath: string;
};

/**
 * Apply autofix for input code
 * */
export const applyEslintAutofix = async ({
  code,
  eslintConfigPath,
  filePath,
}: ApplyEslintAutoFixParams): Promise<string> => {
  try {
    if (!existsSync(eslintConfigPath)) {
      throw new Error(`ESLint configuration file not found by path: ${eslintConfigPath}`);
    }

    const eslint = new ESLint({
      fix: true,
      overrideConfigFile: eslintConfigPath,
    });

    const results = await eslint.lintText(code, { filePath });

    await ESLint.outputFixes(results);

    return results[0]?.output || code;
  } catch (error) {
    console.error('Error applying eslint autofix:', error);
    return code;
  }
};

export const applyEslint = async (code: string) =>
  await applyEslintAutofix({ code, filePath: 'temp/forEslintAutofix.ts', eslintConfigPath: 'eslint.config.mjs' });
