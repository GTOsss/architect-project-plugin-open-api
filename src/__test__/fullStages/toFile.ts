import { toArrayTypes, toArrayMethods } from '#/transformJsonDocument';
import { readYaml } from '#/__test__/utils/fs';
import { resolve } from 'path';
import { OpenAPIV3 } from 'openapi-types';
import { TypesMap } from '#/transformJsonDocument/types';
import { genCode, requestAxiosTemplate } from '#/__test__/fullStages/templates';
import upperFirst from 'lodash.upperfirst';

const openApiMock = readYaml(resolve(__dirname, '../../mock/openapi.mock.yaml')) as OpenAPIV3.Document;

const typesMap: TypesMap = {
  number: 'number',
  integer: 'number',
  string: 'string',
  boolean: 'boolean',
  anyObject: 'AnyObject', // если в scheme только type: object
  '*': 'any', // any если не удалось определить тип
};

const config = { typesMap };

export const toTypescriptFile = () => {
  const methods = toArrayMethods(openApiMock, config);

  return methods
    .map(({ method, parametersMap, responseType, requestBodyType }) => {
      return genCode(requestAxiosTemplate, {
        methodType: method,
        methodTypeUppercase: upperFirst(method),
        responseType,
        urlParams: 'urlparams',
        queryParams: 'queryparams',
        methodName: 'methodName',
        bodyType: requestBodyType,
        methodNameUppercase: 'MethodNameUppercase',
        urlGetter: 'urlgeeter',
      });
    })
    .join('\n ////////// \n');
};
