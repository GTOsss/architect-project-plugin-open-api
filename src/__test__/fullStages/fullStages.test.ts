import { readYaml } from '../utils/fs';
import { resolve } from 'path';
import { OpenAPIV3 } from 'openapi-types';
import { toTypescriptFile } from './toFile';
import { applyEslint } from '../utils/code';

describe('Full stages (Open API spec to code file)', () => {
  it('toCodeFile', async () => {
    const code = toTypescriptFile();
    expect(await applyEslint(code)).toMatchSnapshot();
  });
});
