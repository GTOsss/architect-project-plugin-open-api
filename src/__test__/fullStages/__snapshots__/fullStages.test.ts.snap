// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Full stages (Open API spec to code file) toCodeFile 1`] = `
"
export type Post~methodNameUppercase~ = {
  urlParams: urlparams,
  queryParams: queryparams,
  body: { email: string,name: string,role: "user" | "guest" },
}

export const post~methodName~ = ({
  urlParams,
  queryParams,
  body,
}: ~methodTypeUppercaseundefined): User => {
  return axios.post(urlgeeter, body)
}
 ////////// 

export type Get~methodNameUppercase~ = {
  urlParams: urlparams,
  queryParams: queryparams,
  body: undefined,
}

export const get~methodName~ = ({
  urlParams,
  queryParams,
  body,
}: ~methodTypeUppercaseundefined): Array<User> => {
  return axios.post(urlgeeter, body)
}
 ////////// 

export type Put~methodNameUppercase~ = {
  urlParams: urlparams,
  queryParams: queryparams,
  body: User,
}

export const put~methodName~ = ({
  urlParams,
  queryParams,
  body,
}: ~methodTypeUppercaseundefined): User => {
  return axios.post(urlgeeter, body)
}
 ////////// 

export type Patch~methodNameUppercase~ = {
  urlParams: urlparams,
  queryParams: queryparams,
  body: { name: string,role: "admin" | "user" | "guest" },
}

export const patch~methodName~ = ({
  urlParams,
  queryParams,
  body,
}: ~methodTypeUppercaseundefined): User => {
  return axios.post(urlgeeter, body)
}
 ////////// 

export type Get~methodNameUppercase~ = {
  urlParams: urlparams,
  queryParams: queryparams,
  body: undefined,
}

export const get~methodName~ = ({
  urlParams,
  queryParams,
  body,
}: ~methodTypeUppercaseundefined): User => {
  return axios.post(urlgeeter, body)
}
 ////////// 

export type Delete~methodNameUppercase~ = {
  urlParams: urlparams,
  queryParams: queryparams,
  body: undefined,
}

export const delete~methodName~ = ({
  urlParams,
  queryParams,
  body,
}: ~methodTypeUppercaseundefined): undefined => {
  return axios.post(urlgeeter, body)
}
 ////////// 

export type Post~methodNameUppercase~ = {
  urlParams: urlparams,
  queryParams: queryparams,
  body: Array<User>,
}

export const post~methodName~ = ({
  urlParams,
  queryParams,
  body,
}: ~methodTypeUppercaseundefined): Array<User> => {
  return axios.post(urlgeeter, body)
}"
`;
